Add  timestamp to agent commands

jcvenegas opened the pull request:
- Agent now can get log level configuration from cmdline.
- Changed logrus time format to : `TimestampFormat: time.RFC3339Nano` to get nanoseconds precision.
- Start and end of commands is logged with the message:   `{event}_start` or  `{event}_end`
- An special case is logged `agent_started seconds.nanoseconds` this get uptime with good clock precision.  
- Logrus format is JSON for easy parsing

jcvenegas committed:
log: Get loglevel from kernel cmdline

This patch adds support to read kernel cmdline to get
the log level.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas committed:
log: Use Error format for errors

Use error format for errors.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas committed:
log: Show timestamp before and after run a command

This patch adds timestamps to track times for each command

Fixes: #67

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

sboeuf commented on agent.go:
```diff
@@ -1210,3 +1217,42 @@ func winsizeCb(pod *pod, data []byte) error {
+	}
+
+	if logLevel == "" {
+		agentLog.SetLevel(logrus.InfoLevel)
```
You should use `defaultLogLevel` that you defined earlier.

sboeuf commented on agent.go:
```diff
@@ -1210,3 +1217,42 @@ func winsizeCb(pod *pod, data []byte) error {
+
+	level, err := logrus.ParseLevel(logLevel)
+	if err != nil {
+		agentLog.SetLevel(logrus.InfoLevel)
```
Ditto.

sboeuf commented on agent.go:
```diff
@@ -221,7 +221,7 @@ func (p *pod) controlLoop(wg *sync.WaitGroup) {
 			}
 
-			agentLog.Infof("Read on ctl channel failed: %v", err)
+			agentLog.Errorf("Read on ctl channel failed: %v", err)
```
Well that's not an error, it's about the proxy disconnecting, in order to end the agent.

jcvenegas replied:
I think the only way the proxy disconnecting is doing DESTROYPOD ?

sboeuf replied:
Ok fine.

sboeuf commented on network.go:
```diff
@@ -137,7 +137,7 @@ func setupInterfaces(netHandle *netlink.Handle, ifaces []hyper.NetIface) error {
 		}
 
-		agentLog.Infof("Link found %+v", link)
+		agentLog.Warnf("Link found %+v", link)
```
That's an info, not a warning.

sboeuf commented on time.go:
```diff
@@ -0,0 +1,14 @@
+package main
```
Add the Intel license header please.

sboeuf commented on agent_test.go:
```diff
@@ -39,3 +41,38 @@ func TestMain(m *testing.M) {
+		{"Log config debug",
+			args{"agent.debug=debug "}, false},
+		{"Log config warn",
+			args{" systemd kernel=someting agent.debug=warn "}, false},
```
s/kernel=someting/kernel=foo/

sboeuf commented on agent_test.go:
```diff
@@ -39,3 +41,38 @@ func TestMain(m *testing.M) {
+			args{"agent.debug= debug "}, false},
+		{"Log config empty",
+			args{"agent.debug="}, false},
+	}
```
You could also test in case a wrong parameter is given. Something like:
`agent.debug=ino`
`agent.debug=wran`
`agent.debug=degbu`

sboeuf commented on agent_test.go:
```diff
@@ -39,3 +41,38 @@ func TestMain(m *testing.M) {
+				t.Errorf("parseKernelCmdline() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			agentLog.Infof("info message")
+			agentLog.Debugf("debug message")
```
Those messages don't help in the test validation, do they ?
I think we need something a bit more complex. You could create a pipe with `os.Pipe()`, and give one end of the pipe to logrus, and listen on the other end. That way for instance, you could check that you receive `info message` only when you enable at least Info level.

jcvenegas replied:
Indeed, it does not help, I added to help me know if was working but agree I can add the Pipe sugguestion

sboeuf commented on agent.go:
```diff
@@ -1241,13 +1241,13 @@ func parseKernelCmdline(cmdline string) error {
 
 	if logLevel == "" {
-		agentLog.SetLevel(logrus.InfoLevel)
+		agentLog.SetLevel(defaultLogLevel)
```
Addressed in the wrong commit.

sboeuf commented on agent.go:
```diff
@@ -1241,13 +1241,13 @@ func parseKernelCmdline(cmdline string) error {
 	level, err := logrus.ParseLevel(logLevel)
 	if err != nil {
-		agentLog.SetLevel(logrus.InfoLevel)
+		agentLog.SetLevel(defaultLogLevel)
```
Ditto.

sboeuf commented on agent.go:
```diff
@@ -171,6 +171,13 @@ func main() {
 	agentLog.Infof("Agent version: %s", Version)
+
+	if uptime, err := newEventTime(agentStartedEvent); err != nil {
+		agentLog.Errorf("Failed to get uptime %v", err)
```
Use a constant for `"Failed to get uptime"` because you reuse it later.

sboeuf commented on agent.go:
```diff
@@ -799,7 +806,20 @@ func (p *pod) runCmd(cmd hyper.HyperCmd, data []byte) error {
+	if t, err := newEventTime(hyper.CmdToString(cmd) + "_start"); err == nil {
+		agentLog.Infof("%s", t)
+	} else {
+		agentLog.Errorf("Failed to get time:%v", err)
```
Use a constant for the string.

sboeuf commented on agent.go:
```diff
@@ -799,7 +806,20 @@ func (p *pod) runCmd(cmd hyper.HyperCmd, data []byte) error {
+	if t, err := newEventTime(hyper.CmdToString(cmd) + "_end"); err == nil {
+		agentLog.Infof("%s", t)
+	} else {
+		agentLog.Errorf("Failed to get time:%v", err)
```
Ditto

sboeuf commented on time_test.go:
```diff
@@ -0,0 +1,57 @@
+package main
```
Add the Intel license header please.

jodh-intel commented on time.go:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	if err := unix.ClockGettime(unix.CLOCK_MONOTONIC_RAW, &ts); err != nil {
+		return eventTime{Event: "error", Time: ts}, fmt.Errorf("%v", err)
```
This looks odd to me for 2 reasons:

- conventially, if a function return an `error`, you should ignore any other returned values, but you are creating a special "error" event.
- why can't the returned error be just `, err` rather than `, fmt.Errorf("%v", err)` ?

jcvenegas replied:
you are right!, fixed

jodh-intel commented on agent.go:
```diff
@@ -792,7 +806,20 @@ func (p *pod) runCmd(cmd hyper.HyperCmd, data []byte) error {
+	if t, err := newEventTime(hyper.CmdToString(cmd) + "_start"); err == nil {
+		agentLog.Infof("%s", t)
+	} else {
+		agentLog.Errorf("Failed to get time:%v", err)
```
It looks like the returned `error` won't include the event passed to `newEventTime()`. I'd create the string first:

```
event := hyper.CmdToString(cmd) + "_start"

if t, err := newEventTime(event); err == nil {
    agentLog.Infof("%s", t)
} else {
    agentLog.Errorf("Failed to get time for event %v: %v", event, err)
```

jodh-intel commented on agent.go:
```diff
@@ -1210,3 +1237,42 @@ func winsizeCb(pod *pod, data []byte) error {
+	words := strings.Fields(cmdline)
+	for _, w := range words {
+		if strings.HasPrefix(string(w), "agent.debug=") {
+			logLevel = strings.Split(string(w), "agent.debug=")[1]
```
I'd create a variable rather than calling `string()` twice:

```
str := string(w)
```

Also, I think you should check the number of fields returned by `Split()` before de-referencing the 2nd element.

MarioCarrilloA commented on agent.go:
```diff
@@ -1210,3 +1241,47 @@ func winsizeCb(pod *pod, data []byte) error {
 	return nil
 }
+
+// Get the agent configuration from kenrel cmdline
```
s/kenrel/kernel

sboeuf commented on agent.go:
```diff
@@ -59,6 +59,9 @@ const (
 	statelessPassword  = "/usr/share/defaults/etc/passwd"
```
Please edit the commit message and replace: `s/kenel/kernel/`

sboeuf commented on agent.go:
```diff
@@ -1210,3 +1222,51 @@ func winsizeCb(pod *pod, data []byte) error {
+	words := strings.Fields(cmdline)
+
+	for _, w := range words {
+		for o := range agentOptions {
```
I am sorry but I don't like that we loop on agentOptions and at the same time we are modifying its values. Try something  more simple to stay as clear as possible:
I think we should expect `getConfig()` to return a config, and for each possible option parsed by `parseKernelCmdLine()`, we want a default in case it is not provided by the command line.
This means that `parseKernelCmdLine()` will build a list of options, that is returned to `getConfig()`, and from the `main()`, we apply this list of options in any case because we know that we will always have some default values (a function like `applyConfig()` would be fine). No need to specify any error, maybe just a `log.Warn()` in case we could read `/proc/cmdline` for example.

Moreover, about the way you loop over every element of this, because this is a map, you can directly have `key/value` from the loop for:
```go
for key, value := range optionList {
    ...
}
```

jcvenegas replied:
Hi @sboeuf , I updated the PR:
1. I created agentConfig struct (by now we only have logLevel)
2. getConfig() return parse the kernel cmdline 
 - If could not parse an option I will use a default option.
3. applyConfig(config) will take a config and apply the values already parsed and validated in getConfig()

sboeuf commented on agent.go:
```diff
@@ -1261,7 +1261,7 @@ func parseKernelCmdline(cmdline string) error {
 	level, err := logrus.ParseLevel(agentOptions[logLevelFlag])
 	if err != nil {
-		agentLog.SetLevel(logrus.InfoLevel)
+		agentLog.SetLevel(defaultLogLevel)
```
This should be part of the previous commit.

sboeuf commented on agent.go:
```diff
@@ -62,6 +62,7 @@ const (
 	kernelCmdlineFile  = "/proc/cmdline"
```
This commit message needs to be edited. `s/path/patch/` and `s/add/adds/`

grahamwhaley commented on time.go:
```diff
@@ -0,0 +1,60 @@
+//Event that will be reported for time tracing
+type eventTime struct {
+	Event string
+	Time  unix.Timespec
```
Hi @jcvenegas You asked me what time format I thought we should print in. My gut feeling is that we should probably use the same format as the other logs we produce within CC - so, for instance, it looks like the runtime logs use the logrus internal Time data:
https://github.com/clearcontainers/runtime/blob/9be2f54966b90014251afa5cebdf1ee36a85503f/logger.go#L137
If we can, I think we should do the same - what do you think @jodh-intel

jodh-intel replied:
The actual timestamp itself should be in https://en.wikipedia.org/wiki/ISO_8601 format, or the RFC 3339 variant used by https://golang.org/pkg/time/. Looks like logrus uses that by default.

sboeuf commented on agent.go:
```diff
@@ -1210,3 +1221,59 @@ func winsizeCb(pod *pod, data []byte) error {
+		return nil
+	}
+
+	option = split[optionPosition]
```
Don't assign the key that you parsed to the option, instead, use the key (that is `split[optionPosition]`) directly into the switch:
```go
switch split[optionPosition] {
    ...
```

sboeuf commented on agent.go:
```diff
@@ -1210,3 +1221,59 @@ func winsizeCb(pod *pod, data []byte) error {
+		if err != nil {
+			return err
+		}
+		c.logLevel = level
```
You could add a default case here, and return an error like:
```go
default:
    return fmt.Errorf("Unknown option %s", split[optionPosition])
```

sboeuf reviewed (commented):
Two comments, but other than that, the PR looks good to me now :)

jodh-intel commented on agent.go:
```diff
@@ -1210,3 +1244,59 @@ func winsizeCb(pod *pod, data []byte) error {
+func getConfig() agentConfig {
+
+	config := agentConfig{}
+	config.logLevel = defaultLogLevel
```
I think it would be better if this function also returned an `error`. The caller can then set `config.logLevel = defaultLogLevel`. It's neat to encapsulate the default handling like this, but it makes testing it difficult (how can you know which codepath was followed if `kernelCmdlineFile` contained the same `defaultLogLevel` value?)

Speaking of which, can you add a unit test for this function :smile:

sboeuf commented:
LGTM
Let's wait for @jodh-intel comments before to merge it.

jcvenegas commented:
@jodh-intel I still need to rework the timestamp format...

jcvenegas commented:
@jodh-intel plase take a look to the log using logrus timestamp:
```
{"level":"info","msg":"Agent version: 0.1.0-alpha.0+-28cad9a9947ceced9a05b43ac3db5f4358ea5167","time":"2017-08-15T14:43:11.690803064Z"}
{"level":"info","msg":"agent_started 0.542392826","time":"2017-08-15T14:43:11.692726511Z"}
{"level":"info","msg":"Received \"startpod\" command","time":"2017-08-15T14:43:11.751076463Z"}
{"level":"info","msg":"startpod_start","time":"2017-08-15T14:43:11.751832257Z"}
{"level":"info","msg":"startpod_end","time":"2017-08-15T14:43:11.759136667Z"}
{"level":"info","msg":"Received \"newcontainer\" command","time":"2017-08-15T14:43:11.767398068Z"}
{"level":"info","msg":"newcontainer_start","time":"2017-08-15T14:43:11.767548997Z"}
```

jodh-intel commented:
@jcvenegas - I'm not clear on your last comment. This looks fine to me.

However, are you planning to either modify the logrus formatter to change the time format, or maybe add a special time format *message* for `startpod_start` and `startpod_end` or something?

jcvenegas commented:
@jodh-intel @grahamwhaley  I would like to get your approval in the format, now that I moved to RFC3339 as was suggested. 
- Yes I changed logrus time format to : `TimestampFormat: time.RFC3339Nano` to get nanoseconds precision.
- To log start end events I just print  `{event}_start` or  `{event}_end`
- An special case is logged `agent_started seconds.nanoseconds` this get uptime with good clock precision.  
- Logrus format is JSON for easy parsing

sboeuf commented:
@jodh-intel @grahamwhaley could you give us your review on this one ?

jodh-intel commented:
lgtm.

[![Approved with PullApprove](https://img.shields.io/badge/reviewers-approved-brightgreen.svg)](https://pullapprove.com/clearcontainers/agent/pull-request/70/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent)

grahamwhaley commented:
I'm happy if @jodh-intel is happy, and he is - so merging....
